File: backend/app/services/chat_service.py
import logging
from collections.abc import AsyncGenerator

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.models.chat import Conversation, Message
from app.services import gigachat_client, vector_store
from app.services.document_service import CONTEXT_CHAR_LIMIT

logger = logging.getLogger(__name__)


async def get_conversations(db: AsyncSession, search: str = "") -> list[Conversation]:
    query = select(Conversation).order_by(Conversation.updated_at.desc())
    if search:
        query = query.where(Conversation.title.ilike(f"%{search}%"))
    result = await db.execute(query)
    return list(result.scalars().all())


async def get_conversation(db: AsyncSession, conversation_id: str) -> Conversation | None:
    result = await db.execute(
        select(Conversation)
        .where(Conversation.id == conversation_id)
        .options(selectinload(Conversation.messages))
    )
    return result.scalar_one_or_none()


async def create_conversation(
    db: AsyncSession, title: str | None = None, knowledge_base_id: str | None = None
) -> Conversation:
    conversation = Conversation(
        title=title or "Новый диалог",
        knowledge_base_id=knowledge_base_id,
    )
    db.add(conversation)
    await db.commit()
    await db.refresh(conversation)
    return conversation


async def delete_conversation(db: AsyncSession, conversation_id: str) -> bool:
    conversation = await get_conversation(db, conversation_id)
    if not conversation:
        return False
    await db.delete(conversation)
    await db.commit()
    return True


def _build_documents_context(
    document_ids: list[str],
    document_texts: list[str],
    query: str,
) -> str:
    """Build system context from one or more documents.

    Strategy: process each document individually.
    - Small docs (< CONTEXT_CHAR_LIMIT) are included in full.
    - Large docs get similarity search with more chunks.
    This ensures small documents (like a KP) are never lost,
    while large ones (like a PZ) are searched for relevant parts.
    """
    combined_text = "\n\n".join(document_texts)
    total_len = len(combined_text)

    if total_len <= CONTEXT_CHAR_LIMIT:
        # All documents fit — include everything
        header = (
            f"Пользователь загрузил {len(document_texts)} документ(ов). "
            "Используй их содержимое для ответа на вопросы.\n\n"
        )
        parts = []
        for i, text in enumerate(document_texts):
            parts.append(f"--- ДОКУМЕНТ {i + 1} ---\n{text}\n--- КОНЕЦ ДОКУМЕНТА {i + 1} ---")
        return header + "\n\n".join(parts)

    # Mixed strategy: small docs in full, large docs via similarity search
    parts: list[str] = []
    budget = CONTEXT_CHAR_LIMIT
    query_embedding = None

    for i, (doc_id, text) in enumerate(zip(document_ids, document_texts)):
        doc_num = i + 1

        if len(text) <= budget:
            # Document fits in remaining budget — include in full
            parts.append(
                f"--- ДОКУМЕНТ {doc_num} (полный текст) ---\n{text}\n"
                f"--- КОНЕЦ ДОКУМЕНТА {doc_num} ---"
            )
            budget -= len(text)
        else:
            # Document too large — similarity search
            chunks: list[str] = []
            try:
                if query_embedding is None:
                    query_embedding = gigachat_client.get_embeddings([query])[0]
                chunks = vector_store.search_similar(doc_id, query_embedding, top_k=7)
            except Exception:
                logger.exception("Similarity search failed for document %s", doc_id)

            if chunks:
                joined = "\n\n".join(chunks)
                # Only use what fits in budget
                if len(joined) > budget:
                    joined = joined[:budget]
                parts.append(
                    f"--- ДОКУМЕНТ {doc_num} (релевантные фрагменты) ---\n{joined}\n"
                    f"--- КОНЕЦ ДОКУМЕНТА {doc_num} ---"
                )
                budget -= len(joined)
            else:
                # Fallback: truncate to remaining budget
                truncated = text[:max(budget, 2000)]
                parts.append(
                    f"--- ДОКУМЕНТ {doc_num} (обрезан) ---\n{truncated}\n"
                    f"--- КОНЕЦ ДОКУМЕНТА {doc_num} ---"
                )
                budget -= len(truncated)

    header = (
        f"Пользователь загрузил {len(document_texts)} документ(ов). "
        "Используй их содержимое для ответа на вопросы.\n\n"
    )
    return header + "\n\n".join(parts)


def _build_rag_context(kb_id: str, query: str) -> str | None:
    """Search knowledge base and build context from relevant chunks."""
    try:
        query_embedding = gigachat_client.get_embeddings([query])[0]
        chunks = vector_store.kb_search_similar(kb_id, query_embedding, top_k=5)
        if not chunks:
            return None
        logger.info("RAG: found %d chunks in KB %s", len(chunks), kb_id)
        joined = "\n\n---\n\n".join(chunks)
        return (
            "Ты — ассистент с доступом к базе знаний. Используй приведённые ниже "
            "фрагменты для ответа на вопрос пользователя. Если информации недостаточно, "
            "скажи об этом.\n\n"
            "--- ФРАГМЕНТЫ ИЗ БАЗЫ ЗНАНИЙ ---\n"
            f"{joined}\n"
            "--- КОНЕЦ ФРАГМЕНТОВ ---"
        )
    except Exception:
        logger.exception("RAG search failed for KB %s", kb_id)
        return None


async def send_message_stream(
    db: AsyncSession,
    conversation_id: str,
    content: str,
    document_ids: list[str] | None = None,
    document_texts: list[str] | None = None,
) -> AsyncGenerator[str, None]:
    """Save user message, stream GigaChat response, save assistant message."""
    conversation = await get_conversation(db, conversation_id)
    if not conversation:
        raise ValueError("Conversation not found")

    # Save user message
    user_message = Message(
        conversation_id=conversation_id,
        role="user",
        content=content,
        document_ids=document_ids or [],
    )
    db.add(user_message)
    await db.commit()

    # Build message history for GigaChat
    messages: list[dict[str, str]] = []

    # If documents attached to this message, add as system context
    if document_ids and document_texts:
        messages.append({
            "role": "system",
            "content": _build_documents_context(document_ids, document_texts, content),
        })
    # If conversation is linked to a knowledge base, use RAG
    elif conversation.knowledge_base_id:
        rag_context = _build_rag_context(conversation.knowledge_base_id, content)
        if rag_context:
            messages.append({"role": "system", "content": rag_context})

    # Add conversation history
    for msg in conversation.messages:
        messages.append({"role": msg.role, "content": msg.content})

    messages.append({"role": "user", "content": content})

    # Stream response
    full_response = ""
    usage_info = None
    async for chunk in gigachat_client.chat_completion_stream(messages):
        if isinstance(chunk, dict):
            usage_info = chunk
        else:
            full_response += chunk
            yield chunk

    # Yield usage info as a special marker
    if usage_info:
        yield usage_info

    # Save assistant message
    assistant_message = Message(
        conversation_id=conversation_id,
        role="assistant",
        content=full_response,
    )
    db.add(assistant_message)
    await db.commit()

    # Auto-generate title after first message exchange
    if conversation.title == "Новый диалог" and len(conversation.messages) <= 1:
        try:
            title_response = gigachat_client.chat_completion([
                {
                    "role": "system",
                    "content": "Придумай короткий заголовок (3-5 слов) для диалога на основе первого сообщения пользователя. Верни только заголовок, без кавычек и пояснений.",
                },
                {"role": "user", "content": content},
            ])
            new_title = title_response.strip().strip('"').strip("'")[:100]
            if new_title:
                conversation.title = new_title
                await db.commit()
        except Exception:
            logger.exception("Failed to auto-generate title")

File: backend-java/src/main/java/com/gigame/service/ChatService.java
package com.gigame.service;

import com.gigame.dto.chat.ConversationResponse;
import com.gigame.dto.chat.ConversationWithMessages;
import com.gigame.dto.chat.MessageResponse;
import com.gigame.model.Conversation;
import com.gigame.model.Document;
import com.gigame.model.Message;
import com.gigame.repository.ConversationRepository;
import com.gigame.repository.DocumentRepository;
import com.gigame.repository.MessageRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.codec.ServerSentEvent;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import reactor.core.scheduler.Schedulers;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

/**
 * Chat service — port of Python chat_service.py.
 * Manages conversations, messages, and streaming chat completions with
 * optional document context and RAG from knowledge bases.
 */
@Service
public class ChatService {

    private static final Logger log = LoggerFactory.getLogger(ChatService.class);

    /**
     * Documents shorter than this are included as full text in the prompt.
     * Larger documents use similarity search for relevant chunks.
     */
    private static final int CONTEXT_CHAR_LIMIT = 12_000;

    /**
     * Maximum number of conversation history messages to include in the prompt.
     * Older messages are dropped to keep the context window manageable.
     */
    private static final int MAX_HISTORY_MESSAGES = 20;

    private final ConversationRepository conversationRepository;
    private final MessageRepository messageRepository;
    private final DocumentRepository documentRepository;
    private final GigaChatClient gigaChatClient;
    private final EmbeddingService embeddingService;
    private final DocumentService documentService;

    public ChatService(ConversationRepository conversationRepository,
                       MessageRepository messageRepository,
                       DocumentRepository documentRepository,
                       GigaChatClient gigaChatClient,
                       EmbeddingService embeddingService,
                       DocumentService documentService) {
        this.conversationRepository = conversationRepository;
        this.messageRepository = messageRepository;
        this.documentRepository = documentRepository;
        this.gigaChatClient = gigaChatClient;
        this.embeddingService = embeddingService;
        this.documentService = documentService;
    }

    // -------------------------------------------------------------------------
    // Conversations CRUD
    // -------------------------------------------------------------------------

    /**
     * Return all conversations, optionally filtered by title substring.
     */
    public List<ConversationResponse> getConversations(String search) {
        List<Conversation> conversations;
        if (search != null && !search.isBlank()) {
            conversations = conversationRepository.findByTitleContainingIgnoreCaseOrderByUpdatedAtDesc(search);
        } else {
            conversations = conversationRepository.findAllByOrderByUpdatedAtDesc();
        }
        return conversations.stream().map(this::toConversationResponse).toList();
    }

    /**
     * Return a single conversation with its messages. Throws if not found.
     */
    @org.springframework.transaction.annotation.Transactional(readOnly = true)
    public ConversationWithMessages getConversation(UUID id) {
        Conversation conv = conversationRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Conversation not found: " + id));
        // Use direct query to avoid LazyInitializationException (open-in-view=false)
        List<MessageResponse> msgs = messageRepository.findByConversationIdOrderByCreatedAtAsc(id).stream()
                .map(this::toMessageResponse)
                .toList();
        return new ConversationWithMessages(
                conv.getId(),
                conv.getTitle(),
                conv.getKnowledgeBaseId() != null ? conv.getKnowledgeBaseId().toString() : null,
                conv.getCreatedAt(),
                conv.getUpdatedAt(),
                msgs
        );
    }

    /**
     * Create a new conversation.
     */
    @Transactional
    public ConversationResponse createConversation(String title, String knowledgeBaseId) {
        Conversation conv = new Conversation();
        conv.setTitle(title != null && !title.isBlank() ? title : "Новый диалог");
        if (knowledgeBaseId != null && !knowledgeBaseId.isBlank()) {
            conv.setKnowledgeBaseId(UUID.fromString(knowledgeBaseId));
        }
        conv = conversationRepository.save(conv);
        return toConversationResponse(conv);
    }

    /**
     * Update conversation title.
     */
    @Transactional
    public ConversationResponse updateConversation(UUID id, String title) {
        Conversation conv = conversationRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Conversation not found: " + id));
        if (title != null && !title.isBlank()) {
            conv.setTitle(title);
        }
        conv = conversationRepository.save(conv);
        return toConversationResponse(conv);
    }

    /**
     * Delete a conversation by id. Throws if not found.
     */
    @Transactional
    public void deleteConversation(UUID id) {
        Conversation conv = conversationRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Conversation not found: " + id));
        conversationRepository.delete(conv);
    }

    // -------------------------------------------------------------------------
    // Streaming chat
    // -------------------------------------------------------------------------

    /**
     * Send a user message and stream the assistant response as SSE events.
     *
     * <p>Flow:
     * <ol>
     *   <li>Load conversation (throw if missing)</li>
     *   <li>Save user message to DB</li>
     *   <li>Build GigaChat messages list (system context + history + user message)</li>
     *   <li>Stream response from GigaChat</li>
     *   <li>Save assistant message on completion</li>
     *   <li>Emit SSE: {content: chunk}, {usage: {...}}, {done: true}</li>
     * </ol>
     *
     * @param conversationId the conversation UUID
     * @param content        the user message text
     * @param documentIds    optional list of document UUIDs attached to this message
     * @param documentTexts  optional inline document texts (from upload)
     * @return Flux of SSE events
     */
    public Flux<ServerSentEvent<Map<String, Object>>> sendMessageStream(
            UUID conversationId,
            String content,
            List<String> documentIds,
            List<String> documentTexts) {

        // All blocking operations (JPA, GigaChat token refresh) must run
        // on boundedElastic, not the Netty event loop.
        // All blocking ops (JPA, token refresh) on boundedElastic
        return Flux.defer(() -> {
            Conversation conversation = conversationRepository.findById(conversationId)
                    .orElseThrow(() -> new IllegalArgumentException("Conversation not found: " + conversationId));

            Message userMessage = new Message();
            userMessage.setConversationId(conversationId);
            userMessage.setRole("user");
            userMessage.setContent(content);
            userMessage.setDocumentIds(documentIds);
            messageRepository.save(userMessage);

            List<Map<String, String>> chatMessages = buildChatMessages(conversation, content, documentIds, documentTexts);
            StringBuilder fullResponse = new StringBuilder();

            return gigaChatClient.chatCompletionStream(chatMessages)
                    .map(chunk -> {
                        fullResponse.append(chunk);
                        Map<String, Object> event = new HashMap<>();
                        event.put("content", chunk);
                        return ServerSentEvent.<Map<String, Object>>builder()
                                .data(event).build();
                    })
                    .concatWith(Mono.fromCallable(() -> {
                        saveAssistantMessage(conversationId, fullResponse.toString());
                        autoGenerateTitle(conversation, content);
                        Map<String, Object> doneEvent = new HashMap<>();
                        doneEvent.put("done", true);
                        return ServerSentEvent.<Map<String, Object>>builder()
                                .data(doneEvent).build();
                    }));
        }).subscribeOn(Schedulers.boundedElastic());
    }

    // -------------------------------------------------------------------------
    // Message building
    // -------------------------------------------------------------------------

    /**
     * Build the full messages list for GigaChat including system context,
     * conversation history, and the current user message.
     */
    private List<Map<String, String>> buildChatMessages(
            Conversation conversation,
            String userContent,
            List<String> documentIds,
            List<String> documentTexts) {

        List<Map<String, String>> messages = new ArrayList<>();

        // System context: attached documents or RAG
        if (documentTexts != null && !documentTexts.isEmpty()) {
            List<String> ids = documentIds != null ? documentIds : List.of();
            String docContext = buildDocumentsContext(ids, documentTexts, userContent);
            messages.add(Map.of("role", "system", "content", docContext));
        } else if (conversation.getKnowledgeBaseId() != null) {
            String ragContext = buildRagContext(conversation.getKnowledgeBaseId(), userContent);
            if (ragContext != null) {
                messages.add(Map.of("role", "system", "content", ragContext));
            }
        }

        // Conversation history (sliding window: keep only last MAX_HISTORY_MESSAGES)
        List<Message> history = messageRepository.findByConversationIdOrderByCreatedAtAsc(conversation.getId());
        if (history.size() > MAX_HISTORY_MESSAGES) {
            history = history.subList(history.size() - MAX_HISTORY_MESSAGES, history.size());
        }
        for (Message msg : history) {
            messages.add(Map.of("role", msg.getRole(), "content", msg.getContent()));
        }

        // Current user message
        messages.add(Map.of("role", "user", "content", userContent));

        return messages;
    }

    /**
     * Build system context from attached documents.
     *
     * <p>Strategy: small documents (&lt; CONTEXT_CHAR_LIMIT total) are included in full.
     * Large documents get similarity search for relevant chunks.
     */
    private String buildDocumentsContext(List<String> documentIds, List<String> documentTexts, String query) {
        String combined = String.join("\n\n", documentTexts);
        int totalLen = combined.length();

        if (totalLen <= CONTEXT_CHAR_LIMIT) {
            // All documents fit — include everything
            StringBuilder sb = new StringBuilder();
            sb.append("Пользователь загрузил ").append(documentTexts.size())
                    .append(" документ(ов). Используй их содержимое для ответа на вопросы.\n\n");
            for (int i = 0; i < documentTexts.size(); i++) {
                sb.append("--- ДОКУМЕНТ ").append(i + 1).append(" ---\n")
                        .append(documentTexts.get(i))
                        .append("\n--- КОНЕЦ ДОКУМЕНТА ").append(i + 1).append(" ---\n\n");
            }
            return sb.toString();
        }

        // Mixed strategy: small docs in full, large docs via similarity search
        List<String> parts = new ArrayList<>();
        int budget = CONTEXT_CHAR_LIMIT;
        float[] queryEmbedding = null;

        for (int i = 0; i < documentTexts.size(); i++) {
            String docId = i < documentIds.size() ? documentIds.get(i) : null;
            String text = documentTexts.get(i);
            int docNum = i + 1;

            if (text.length() <= budget) {
                parts.add("--- ДОКУМЕНТ " + docNum + " (полный текст) ---\n" + text
                        + "\n--- КОНЕЦ ДОКУМЕНТА " + docNum + " ---");
                budget -= text.length();
            } else {
                // Document too large — similarity search
                List<String> chunks = List.of();
                if (docId != null) {
                    try {
                        if (queryEmbedding == null) {
                            queryEmbedding = gigaChatClient.getEmbeddings(List.of(query)).get(0);
                        }
                        chunks = embeddingService.searchSimilar("doc", UUID.fromString(docId), queryEmbedding, 7);
                    } catch (Exception e) {
                        log.error("Similarity search failed for document {}", docId, e);
                    }
                }

                if (!chunks.isEmpty()) {
                    String joined = String.join("\n\n", chunks);
                    if (joined.length() > budget) {
                        joined = joined.substring(0, budget);
                    }
                    parts.add("--- ДОКУМЕНТ " + docNum + " (релевантные фрагменты) ---\n" + joined
                            + "\n--- КОНЕЦ ДОКУМЕНТА " + docNum + " ---");
                    budget -= joined.length();
                } else {
                    // Fallback: truncate
                    int cutLen = Math.max(budget, 2000);
                    String truncated = text.substring(0, Math.min(cutLen, text.length()));
                    parts.add("--- ДОКУМЕНТ " + docNum + " (обрезан) ---\n" + truncated
                            + "\n--- КОНЕЦ ДОКУМЕНТА " + docNum + " ---");
                    budget -= truncated.length();
                }
            }
        }

        StringBuilder sb = new StringBuilder();
        sb.append("Пользователь загрузил ").append(documentTexts.size())
                .append(" документ(ов). Используй их содержимое для ответа на вопросы.\n\n");
        sb.append(String.join("\n\n", parts));
        return sb.toString();
    }

    /**
     * Search the knowledge base and build RAG system context.
     *
     * @param kbId  the knowledge base UUID
     * @param query the user's query text
     * @return system message content, or null if no relevant chunks found
     */
    private String buildRagContext(UUID kbId, String query) {
        try {
            float[] queryEmbedding = gigaChatClient.getEmbeddings(List.of(query)).get(0);
            List<String> chunks = embeddingService.searchSimilar("kb", kbId, queryEmbedding, 5);
            if (chunks.isEmpty()) {
                return null;
            }
            log.info("RAG: found {} chunks in KB {}", chunks.size(), kbId);
            String joined = String.join("\n\n---\n\n", chunks);
            return "Ты — ассистент с доступом к базе знаний. Используй приведённые ниже "
                    + "фрагменты для ответа на вопрос пользователя. Если информации недостаточно, "
                    + "скажи об этом.\n\n"
                    + "--- ФРАГМЕНТЫ ИЗ БАЗЫ ЗНАНИЙ ---\n"
                    + joined + "\n"
                    + "--- КОНЕЦ ФРАГМЕНТОВ ---";
        } catch (Exception e) {
            log.error("RAG search failed for KB {}", kbId, e);
            return null;
        }
    }

    // -------------------------------------------------------------------------
    // Persistence helpers
    // -------------------------------------------------------------------------

    @Transactional
    protected void saveAssistantMessage(UUID conversationId, String content) {
        Message assistantMessage = new Message();
        assistantMessage.setConversationId(conversationId);
        assistantMessage.setRole("assistant");
        assistantMessage.setContent(content);
        messageRepository.save(assistantMessage);
    }

    /**
     * Auto-generate a conversation title after the first message exchange.
     */
    private void autoGenerateTitle(Conversation conversation, String userContent) {
        if (!"Новый диалог".equals(conversation.getTitle())) {
            return;
        }
        List<Message> msgs = messageRepository.findByConversationIdOrderByCreatedAtAsc(conversation.getId());
        // Only auto-title if this is the first exchange (user + assistant = 2 messages)
        if (msgs.size() > 2) {
            return;
        }
        try {
            String titleResponse = gigaChatClient.chatCompletion(List.of(
                    Map.of("role", "system",
                            "content", "Придумай короткий заголовок (3-5 слов) для диалога на основе первого сообщения пользователя. Верни только заголовок, без кавычек и пояснений."),
                    Map.of("role", "user", "content", userContent)
            ));
            String newTitle = titleResponse.strip();
            // Strip surrounding quotes
            if (newTitle.startsWith("\"") || newTitle.startsWith("'")) {
                newTitle = newTitle.substring(1);
            }
            if (newTitle.endsWith("\"") || newTitle.endsWith("'")) {
                newTitle = newTitle.substring(0, newTitle.length() - 1);
            }
            if (newTitle.length() > 100) {
                newTitle = newTitle.substring(0, 100);
            }
            if (!newTitle.isBlank()) {
                conversation.setTitle(newTitle);
                conversationRepository.save(conversation);
            }
        } catch (Exception e) {
            log.error("Failed to auto-generate title", e);
        }
    }

    // -------------------------------------------------------------------------
    // DTO mapping
    // -------------------------------------------------------------------------

    private ConversationResponse toConversationResponse(Conversation conv) {
        return new ConversationResponse(
                conv.getId(),
                conv.getTitle(),
                conv.getKnowledgeBaseId() != null ? conv.getKnowledgeBaseId().toString() : null,
                conv.getCreatedAt(),
                conv.getUpdatedAt()
        );
    }

    private MessageResponse toMessageResponse(Message msg) {
        return new MessageResponse(
                msg.getId(),
                msg.getConversationId(),
                msg.getRole(),
                msg.getContent(),
                msg.getDocumentIds(),
                msg.getCreatedAt()
        );
    }
}
